walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,97 @@
 
 package com.google.sps;
 
+import java.io.*;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.stream.*;
+
 
 public final class FindMeetingQuery {
+
+  /** Returns a Collection of time ranges when meeting {@code request} can be scheduled in the day of 
+    * events so that all attendees are free 
+    */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      // if the meeting lasts more than a day, there is no solution
+      return new ArrayList<TimeRange>();
+    }
+
+    // array that will represent the number of meetings happening during that minute
+    ArrayList<Integer> meetings = new ArrayList<Integer>();
+    // before processing the events, the number of meetings for each minute is 0
+    meetings.addAll(Collections.nCopies(TimeRange.END_OF_DAY - TimeRange.START_OF_DAY + 1, 0));
+    Collection<String> attendees = request.getAttendees();
+
+    // If at least one requested attendee is participating in the event
+    // update the number of meetings for the start and end time of event
+    events.parallelStream().filter(event -> (event.containsRequestedAttendees(attendees) == true))
+                            .map(Event::getWhen)
+                            .collect(Collectors.toList())
+                            .forEach(eventWhen -> updateNumberOfMeetings(meetings, eventWhen));
+
+    // after all events are processed, for each minute x, meetings[x] will represent the number of
+    // meetings that start at minute x, minus the number of meetings that end right before minute x
```

Comment:
`end right before minute x` or `end at minute x`?

Reply:
The array counts the meetings that `end right before minute x` because the ending value - `when.end()` is exclusive.